File: 229-majority-element-ii/229-majority-element-ii.cpp
class Solution {
public:
    void verify(int n,vector<int>&nums,vector<int>& ans){
        int c=0;
        for(auto num:nums){
            if(num==n)c++;
        }
        
        if(c>nums.size()/3&&find(ans.begin(),ans.end(),n)==ans.end())ans.push_back(n);
    }
    
    vector<int> majorityElement(vector<int>& nums) {
        int num1=-1,num2=-1,c1=0,c2=0;
        vector<int> ans;
        for(auto num:nums){
            if(num1==num)c1++;
            else if(num2==num)c2++;
            else if(c1==0) {num1=num;c1=1;}
            else if(c2==0) {num2=num;c2=1;}
            else{c1--;c2--;}
        }
        
        verify(num1,nums,ans);
        verify(num2,nums,ans);
        return ans;
    }
};

File: 229-majority-element-ii/229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        ans = set()
        n1,n2,ct1,ct2 = -1,-1,0,0
        for i in nums:
            if i==n1: ct1+=1
            elif i==n2: ct2+=1
            elif ct1==0:
                n1 = i
                ct1 = 1
            elif ct2==0:
                n2 = i
                ct2 = 1
            else:
                ct2-=1
                ct1-=1
        def verify(num):
            if nums.count(num)>len(nums)//3:
                ans.add(num)
        verify(n1)
        verify(n2)
        return ans